File: android/app/src/main/kotlin/com/sopro/sopro/logging/LogSanitizer.kt
package com.sopro.sopro.logging

// Mascaramento automático de payloads — comportamento idêntico ao LogSanitizer Flutter.
//
// Duas camadas:
//   Camada 1a — chave sensível → valor inteiro substituído por "[REDACTED]"
//   Camada 1b — chave geográfica → substituído por "[LOCATION]"
//   Camada 2  — conteúdo de string: JWT, Bearer, e-mail, CPF, telefone, URL params
//
// Isenções em modo debug (debugLogging == true):
//   transcript, environment_name, gemini_response, intent, speech_result
//   → necessários para diagnóstico de voz e NLU.
@Suppress("UNCHECKED_CAST")
object LogSanitizer {

    private val debugExemptKeys = setOf(
        "transcript",
        "environment_name",
        "gemini_response",
        "intent",
        "speech_result",
    )

    private val sensitiveKeyPattern = Regex(
        """^(?:authorization|apikey|api_key|token|bearer|password|senha|""" +
            """cookie|secret|private_key|access_key|refresh_token|""" +
            """supabase_key|gemini_key|user_id|uid|account_id|profile_id|person_id)$""",
        RegexOption.IGNORE_CASE,
    )

    // \b evita falsos positivos em "platform" (contém "lat"?) ou "relations".
    private val locationKeyPattern = Regex(
        """\blat(?:itude)?\b|\blon(?:gitude)?\b|\blng\b""",
        RegexOption.IGNORE_CASE,
    )

    private val jwtPattern = Regex("""eyJ[\w\-]+\.eyJ[\w\-]+\.[\w\-]+""")
    private val bearerPattern = Regex("""Bearer\s+\S+""", RegexOption.IGNORE_CASE)
    private val emailPattern = Regex("""\b[\w.\-]+@[\w.\-]+\.\w{2,}\b""")
    private val cpfPattern = Regex("""\b\d{3}\.\d{3}\.\d{3}-\d{2}\b""")
    private val phonePattern = Regex(
        """\b(?:\+55\s?)?(?:\(?\d{2}\)?[\s\-]?)(?:9\s?)?\d{4}[\-\s]?\d{4}\b""",
    )
    private val urlTokenPattern = Regex(
        """([?&](?:token|api_key|apikey|access_token|key|secret|auth|authorization|bearer)=)[^&\s]+""",
        RegexOption.IGNORE_CASE,
    )

    // Ponto de entrada público. Retorna payload original se mascaramento desabilitado.
    fun sanitize(payload: Map<String, Any?>): Map<String, Any?> {
        if (!LoggerConfiguration.enableDataMasking) return payload
        return sanitizeMap(payload)
    }

    private fun sanitizeMap(map: Map<String, Any?>): Map<String, Any?> =
        map.mapValues { (key, value) ->
            val isExempt = LoggerConfiguration.debugLogging &&
                debugExemptKeys.contains(key.lowercase())
            if (isExempt) return@mapValues value

            if (sensitiveKeyPattern.matches(key.lowercase())) return@mapValues "[REDACTED]"
            if (locationKeyPattern.containsMatchIn(key)) return@mapValues "[LOCATION]"

            sanitizeValue(value)
        }

    private fun sanitizeValue(value: Any?): Any? = when (value) {
        is String -> sanitizeString(value)
        is Map<*, *> -> sanitizeMap(value as Map<String, Any?>)
        is List<*> -> value.map { sanitizeValue(it) }
        else -> value
    }

    private fun sanitizeString(value: String): String {
        var result = value
        result = jwtPattern.replace(result, "[JWT]")
        result = bearerPattern.replace(result, "Bearer [REDACTED]")
        result = urlTokenPattern.replace(result) { mr -> "${mr.groupValues[1]}[REDACTED]" }
        result = emailPattern.replace(result, "[EMAIL]")
        result = cpfPattern.replace(result, "[CPF]")
        result = phonePattern.replace(result, "[PHONE]")
        return result
    }
}

File: lib/infrastructure/logging/core/log_sanitizer.dart
import 'logger_configuration.dart';

// Aplica mascaramento automático sobre payloads de log antes de qualquer saída
// (console, Supabase, arquivo, etc.).
//
// ── Estratégia de mascaramento (duas camadas) ─────────────────────────────
//
//   Camada 1 — Por nome de chave:
//     Chaves reconhecidas como sensíveis têm o valor substituído por
//     '[REDACTED]' ou '[LOCATION]', independentemente do conteúdo.
//     Exemplos: "token", "password", "lat", "user_id".
//
//   Camada 2 — Por conteúdo de valor (regex):
//     Strings são varridas por padrões que detectam JWTs, Bearer tokens,
//     e-mails, CPFs, telefones brasileiros e URLs com query params sensíveis.
//     Aplicada recursivamente em Maps aninhados e itens de List.
//
// ── Sanitização recursiva ─────────────────────────────────────────────────
//   _sanitizeMap  → processa cada entrada; chama _sanitizeValue por valor
//   _sanitizeValue → despacha para String, Map ou List
//   _sanitizeString → aplica todos os padrões regex em ordem
//   _sanitizeMap  (recursão em Maps aninhados)
//   _sanitizeValue (recursão em listas heterogêneas)
//
// ── Isenções em modo debug ────────────────────────────────────────────────
//   Quando LoggerConfiguration.debugLogging == true, as chaves abaixo NÃO
//   são mascaradas — necessárias para diagnóstico de voz e NLU:
//     transcript, environment_name, gemini_response, intent, speech_result
//   TODOS os demais campos continuam mascarados mesmo em debug.
class LogSanitizer {
  LogSanitizer._();

  // ── Chaves isentas (somente em debug) ────────────────────────────────────

  static const _debugExemptKeys = {
    'transcript',
    'environment_name',
    'gemini_response',
    'intent',
    'speech_result',
  };

  // ── Padrões de nome de chave sensível ────────────────────────────────────

  // Chaves cujo valor inteiro deve ser ocultado com '[REDACTED]'.
  // Usa full-string match via ^...$ para evitar falsos positivos em nomes
  // compostos (ex.: "custom_token_count" não deve ser mascarado).
  static final _sensitiveKeyPattern = RegExp(
    r'^(?:authorization|apikey|api_key|token|bearer|password|senha|'
    r'cookie|secret|private_key|access_key|refresh_token|'
    r'supabase_key|gemini_key|user_id|uid|account_id|profile_id|person_id)$',
    caseSensitive: false,
  );

  // Chaves que contêm coordenadas geográficas → substituídas por '[LOCATION]'.
  // Word boundaries (\b) evitam falsos positivos em keys como "platform"
  // (contém "lat"?) ou "relations" (contém "lat" via "re-lat-ions").
  // Sem \b, hasMatch("platform") retornaria true incorretamente.
  static final _locationKeyPattern = RegExp(
    r'\blat(?:itude)?\b|\blon(?:gitude)?\b|\blng\b',
    caseSensitive: false,
  );

  // ── Padrões de conteúdo de valor ─────────────────────────────────────────

  // JWT compacto: três segmentos base64url separados por ponto.
  // O prefixo "eyJ" é característico do header JSON codificado.
  static final _jwtPattern = RegExp(
    r'eyJ[\w\-]+\.eyJ[\w\-]+\.[\w\-]+',
  );

  // Bearer token em qualquer header ou valor de string.
  static final _bearerPattern = RegExp(
    r'Bearer\s+\S+',
    caseSensitive: false,
  );

  // Endereço de e-mail (padrão RFC 5321 simplificado).
  static final _emailPattern = RegExp(
    r'\b[\w.\-]+@[\w.\-]+\.\w{2,}\b',
  );

  // CPF no formato canônico XXX.XXX.XXX-XX.
  // Não mascara sequências de 11 dígitos sem pontuação para evitar falsos
  // positivos em IDs, timestamps e contadores.
  static final _cpfPattern = RegExp(
    r'\b\d{3}\.\d{3}\.\d{3}-\d{2}\b',
  );

  // Telefone brasileiro: +55 opcional, DDD entre parênteses opcional,
  // 8 ou 9 dígitos com separadores opcionais.
  static final _phonePattern = RegExp(
    r'\b(?:\+55\s?)?(?:\(?\d{2}\)?[\s\-]?)(?:9\s?)?\d{4}[\-\s]?\d{4}\b',
  );

  // Query parameters sensíveis em URLs.
  // Captura o nome do parâmetro (grupo 1) para reescrever apenas o valor.
  // Exemplo: "?token=abc123"   → "?token=[REDACTED]"
  //          "&api_key=xyz"    → "&api_key=[REDACTED]"
  // Interrompe no próximo separador de query (&), espaço ou fim de linha.
  static final _urlTokenPattern = RegExp(
    r'([?&](?:token|api_key|apikey|access_token|key|secret|auth|authorization|bearer)=)[^&\s]+',
    caseSensitive: false,
  );

  // ── API pública ───────────────────────────────────────────────────────────

  // Aplica sanitização ao payload. Retorna uma cópia sanitizada.
  // Se enableDataMasking == false, retorna o mapa original sem cópia.
  static Map<String, dynamic> sanitize(Map<String, dynamic> payload) {
    if (!LoggerConfiguration.enableDataMasking) return payload;
    return _sanitizeMap(payload);
  }

  // ── Implementação interna ─────────────────────────────────────────────────

  static Map<String, dynamic> _sanitizeMap(Map<String, dynamic> map) {
    return map.map((key, value) {
      // Isenção de diagnóstico: aplica somente quando debugLogging está ativo.
      final isExempt = LoggerConfiguration.debugLogging &&
          _debugExemptKeys.contains(key.toLowerCase());
      if (isExempt) return MapEntry(key, value);

      // Camada 1a: chave identificada como sensível → oculta valor inteiro.
      if (_sensitiveKeyPattern.hasMatch(key.toLowerCase())) {
        return MapEntry(key, '[REDACTED]');
      }

      // Camada 1b: chave de coordenada geográfica → placeholder de localização.
      if (_locationKeyPattern.hasMatch(key)) {
        return MapEntry(key, '[LOCATION]');
      }

      // Camada 2: inspeciona o conteúdo do valor (recursivo para Maps e Lists).
      return MapEntry(key, _sanitizeValue(value));
    });
  }

  // Despacha para o sanitizador correto conforme o tipo do valor.
  // Garante sanitização recursiva em toda a hierarquia do payload.
  static dynamic _sanitizeValue(dynamic value) {
    if (value is String) return _sanitizeString(value);
    if (value is Map<String, dynamic>) return _sanitizeMap(value);
    if (value is Map) {
      // Fallback para Map com tipos dinâmicos (ex.: Map<dynamic, dynamic>).
      return _sanitizeMap(value.cast<String, dynamic>());
    }
    if (value is List) {
      return value.map(_sanitizeValue).toList();
    }
    return value;
  }

  // Aplica todos os padrões regex em ordem sobre um valor string.
  // A ordem importa: JWT e Bearer são detectados antes de padrões menores.
  static String _sanitizeString(String value) {
    var result = value;
    result = result.replaceAll(_jwtPattern, '[JWT]');
    result = result.replaceAll(_bearerPattern, 'Bearer [REDACTED]');
    result = result.replaceAllMapped(
      _urlTokenPattern,
      (m) => '${m.group(1)}[REDACTED]',
    );
    result = result.replaceAll(_emailPattern, '[EMAIL]');
    result = result.replaceAll(_cpfPattern, '[CPF]');
    result = result.replaceAll(_phonePattern, '[PHONE]');
    return result;
  }
}
